[1단계 - 장기] 루키(김준하) 미션 제출합니다.

junhaa opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

--> 

안녕하세요 피케이 😃

장기 미션을 완벽하게 구현하지는 못했지만, 피드백을 받아가면서 수정하려고 해요.

이번 미션 잘부탁드려요!
## 체크 리스트

- [X] 미션의 필수 요구사항을 모두 구현했나요?
- [X] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [X] 애플리케이션이 정상적으로 실행되나요?
- [X] [프롤로그](https://prolog.techcourse.co.kr/studylogs/4137)에 셀프 체크를 작성했나요?
  - [프롤로그](https://prolog.techcourse.co.kr/studylogs/4137)<!-- 작성한 셀프 체크의 링크를 남겨주세요. -->


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [X] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

- 한 메서드에 여러 단계의 들여쓰기(indent)를 사용했습니다.
- 모든 엔티티를 작게 유지하지 못했습니다.

<!-- 이유 작성 -->

  
## 어떤 부분에 집중하여 리뷰해야 할까요?
- 테스트 코드의 리팩터링을 진행하지 못했습니다. 추후 테스트 코드의 부족한 부분을 채우려고 합니다.

### 1️⃣ 코드의 중복 제거
현재 코드에서는 `Movements`라는 클래스를 따로 두어 이동 로직의 중복을 제거하고 있습니다. 하지만 모든 추상 클래스 `Piece`의 자식 클래스들이 모든 `Movements`를 사용하는 것이 아닌, `차`와 `포` 객체는 이동로직이 달라서 `Movements`를 사용하고 있지 않습니다. 최대한 중복 코드를 제거하기 위해 `Movements`를 추가하였는데, 지금과 같은 상황에서 기물들의 움직임에서 중복 코드를 제거하기 위해 억지로 분리한 느낌이 듭니다. 이런 경우 약간의 코드 중복을 제거하기 위해서 특정 기물들만 코드를 분리하는 방법이 나을지 아니면, 각자 다른 객체로 보고  따로 기능을 구현해야 할 지 궁금합니다.

### 2️⃣ `Piece`의 `List<Piece>` 참조
현재 `Piece` 기물 객체가 좌표를 가지고 이동하는 기능을 담당하고 있습니다. 하지만 실제 이동 여부를 판별하기 위해서 현재 보드판의 기물들을 확인해야 하는데, 이 경우 `Board` 클래스에서 관리하고 있는 List<Piece>를 전달하고 있습니다. 현재 상황에서 `Piece` 자체가 이동을 담당해야 하고, `Board`와 `Piece` 두 클래스가 서로 강한 연관관계를 가진다고 생각하여 List<Piece>를 전달하여도 된다고 판단했습니다. 이런 경우에서는 List<Piece>를 Piece에게 전달하는 방법에서 제가 생각하지 못한 취약점이 있을 지 궁금합니다.
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->

junhaa committed:
docs: 요구사항 추가

junhaa committed:
feat: 위치 정보를 저장하는 객체 구현

- 보드판의 범위를 검증하는 기능 구현

jaehyeon2650 committed:
feat: solider 말 생성

- 말 이동 기능 추가

junhaa committed:
feat: 졸의 주변이 비어있을 때 이동 가능 여부를 판단하는 기능 구현

jaehyeon2650 committed:
feat: solider 이동 가능 여부 추가

jaehyeon2650 committed:
feat: 말 객체 생성

junhaa committed:
feat: 사용자의 입력에 맞게 상, 마의 배치를 변경하는 기능 구현

jaehyeon2650 committed:
feat: rightElephant 상 마 위치 구현

jaehyeon2650 committed:
feat: LeftElephant 상 마 위치 구현

jaehyeon2650 committed:
feat: 말 위치 초기화 기능 추가

junhaa committed:
feat: 장기판 초기화 기능 구현

junhaa committed:
feat: 최종 목적지의 좌표를 반환하는 기능 구현

jaehyeon2650 committed:
feat: 중간 이동 지점 반환 기능 추가

junhaa committed:
feat: 최종 목적지로 이동이 가능한지 판단하는 기능 구현

junhaa committed:
feat: 이동 경로 지점을 반환하는 기능 구현

jaehyeon2650 committed:
feat: 졸이 이동할 수 있는지 판단하는 로직 추가

junhaa committed:
feat: 마의 이동 가능 여부를 반환하는 기능 구현

jaehyeon2650 committed:
feat: 코끼리 말이 이동할 수 있는지 판단하는 로직 추가

junhaa committed:
feat: 장군의 이동 가능 여부를 반환하는 기능 구현

junhaa committed:
feat: 선비의 이동 가능 여부를 반환하는 기능 구현

jaehyeon2650 committed:
feat: 전차 말이 이동할 수 있는지 판단하는 로직 추가

junhaa committed:
feat: 포가 이동할 수 있는지 판단하는 로직 추가

jaehyeon2650 committed:
feat: 말을 움직이는 기능 추가

junhaa committed:
feat: 게임 승리 팀을 반환하는 기능 구현

jaehyeon2650 committed:
feat: 승자를 조회하는 기능 추가

junhaa committed:
feat: 게임 초기화 과정 구현

jaehyeon2650 committed:
feat: 장기말 출력 기능 추가

junhaa committed:
feat: 장기 게임 실행 로직 구현

junhaa committed:
fix: 포 이동 오류 해결

jaehyeon2650 committed:
feat: 이름 검증 추가

jaehyeon2650 committed:
refactor: 장기 말 상태 예외 발생 순서 변경

junhaa committed:
feat: 사용자의 좌표 입력 검증 추가

jaehyeon2650 committed:
Merge remote-tracking branch 'origin/step1' into step1

junhaa committed:
refactor: view 내부 메서드 추출

jaehyeon2650 committed:
refactor: JanggiRunner 내부 메서드 추출

junhaa committed:
refactor: 코드 내부 메서드 추출

jaehyeon2650 committed:
feat: 같은 종류의 말인지 확인하는 기능 추가

jaehyeon2650 committed:
refactor: Board 메서드 추출

jaehyeon2650 committed:
refactor: 상 말 패턴 추상클래스로 변경

jaehyeon2650 committed:
refactor: PieceFactory 메서드 분리

jaehyeon2650 committed:
refactor: 움직임 판단 메서드 이동

jaehyeon2650 committed:
refactor: 도착지점에 아군인지 판단하는 메서드 이동

jaehyeon2650 committed:
refactor: Cannon 메서드 추출

junhaa committed:
refactor: 차, 포 내부 메서드 분리

pkeugine commented on src/main/java/domain/JanggiRunner.java:
```diff
@@ -0,0 +1,93 @@
+
+    private void executeGame(JanggiGame janggiGame) {
+        TeamType nowTurn = TeamType.CHO;
+        while (isGameInProgress(janggiGame)) {
```
`isGameInProgress` 메소드를 `JanggiRunner` 에서 추가로 만들어서 호출하지 말고
`JanggiGame` 내부에서 만든 메소드를 바로 호출하면 어떨까요?
불필요한 `depth`는 줄이는 것이 좋습니다.


```suggestion
        while (janggiGame.isGameInProgress()) {
```

junhaa replied:
조건에 대한 부정문을 제거하기 위해 `JanggiRunner` 내부에 `IsGameInProgress()` 메서드를 두었는데 오히려 불필요한 `depth`가 생긴 것 같아요. `JanggiGame` 내부에서 `isInProgress()`메서드를 호출하도록 변경하겠습니다!

pkeugine commented on src/main/java/domain/JanggiGame.java:
```diff
@@ -0,0 +1,36 @@
+    public void movePiece(Position startPosition, Position endPosition, TeamType teamType) {
+        board.movePiece(startPosition, endPosition, teamType);
+    }
+
```
가독성을 위해 전체 코드에서 줄바꿈을 일관성있게 맞춰주세요.

pkeugine commented on src/main/java/domain/Row.java:
`Column` 과 `Row`의 코드가 거의 동일하네요.
공통 부분을 재사용할 수 있도록 개선해보면 어떨까요?

junhaa replied:
`Column` 와 `Row`에서 대부분의 코드가 중복되네요. `Axis`를 추상클래스로 두어 중복 부분을 제거하고 재사용할 수 있도록 변경해보겠습니다!

pkeugine commented on src/main/java/domain/Board.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    public void movePiece(Position startPosition, Position endPosition, TeamType team) {
+        Piece findPiece = findPieceByPosition(startPosition)
```
`findPiece`는 변수보다는 메소드에 어울리는 이름 같아요.

junhaa replied:
변수명을 `동사 + 명사`로 지어버렸네요 🥲 변수명에 맞게 수정해보겠습니다!

pkeugine commented on src/main/java/domain/piece/Elephant.java:
```diff
@@ -0,0 +1,53 @@
+    public boolean canMove(Position expectedPosition, List<Piece> pieces) {
+        if(!MOVEMENTS.canMovePieceToPosition(this, expectedPosition, pieces)){
+            return false;
+        };
```
중괄호 뒤에 `;`은 불필요합니다.

junhaa replied:
코드를 복사하는 과정에서 `;`을 같이 복사해서 전체적인 코드에 `;`가 붙어있네요 🥲 수정하도록 하겠습니다!

pkeugine commented on src/main/java/domain/piece/Piece.java:
기물들이 모두 `Piece`를 상속하고 있음에도 불구하고 중복된 코드가 많습니다.
공통 로직을 `Piece`로 끌어올려 중복을 제거해보면 어떨까요?
특히 `canMove()` 및 관련 `private` 메서드들을 중심으로 개선해보면 좋을듯 합니다. :)

junhaa replied:
아무래도 상속을 사용했음에도 중복된 코드를 제거하지 못한 것 같아요 😭 내부의 공통 로직을 `Piece`로 이동시켜 중복을 제거하는 방법을 고민해보겠습니다!

pkeugine replied:
코드가 정말 많이 개선 됐네요.
훨씬 간결해지고 가독성도 좋아졌습니다 💯

pkeugine commented on src/main/java/domain/Movements.java:
말을 이동시키기 위해 `Movements` → `Movement` → 방향 계산 → 중간 위치 탐색까지 흐름이 깊어져서,
이해하기 쉽지 않아 보입니다.
조금 더 구조를 단순화할 수 있는 방법을 고민해보면 좋을 것 같아요.

junhaa replied:
현재 구조가 말이 이동하는 로직 하나에서 너무 깊은 depth로 일어나서, 코드 자체가 복잡해진 것 같아요. 말의 이동 로직에 대해 여러 방법을 적용해 본 후, 고민해 보도록 하겠습니다!

pkeugine replied:
이 부분도 코드가 정말 많이 개선 됐네요. 👍

pkeugine commented on src/main/java/domain/piece/Cannon.java:
```diff
@@ -0,0 +1,115 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class Cannon extends Piece {
```
`Movements`가 `Chariot`과 `Cannon`에서만 사용되지 않는 이유가 있을까요?

junhaa replied:
현재 `Movements` 클래스는 **고정된 길이만큼 움직이는 기물**에 대한 **이동 경로**를 이용하여 움직임을 검증하고 있습니다. `Chariot`이나 `Cannon` 두 기물은 이동 길이가 고정적이지 않기 때문에 `Movements` 클래스에는 적용하지 못했습니다. 

결국 두 가지 움직임에 대한 로직이 다르기 때문에 `고정 길이 이동` 과 `가변 길이 이동`으로 나누어서 관리하는 방법을 이용하면 중복 코드를 제거 할 수 있을 것 같습니다. 하지만 움직임에 대한 로직을 분리하는 것이 중복 코드 제거를 위해 억지로 분리하는 느낌이 드는데, 제가 분리하는 기준을 잘못 생각하고 있어서 그런 것 인지 궁금합니다.

pkeugine replied:
"고정 길이 이동"과 "가변 길이 이동"은 서로 다른 전략으로 분리될 수 있습니다.
도메인 상의 "이동 규칙"이 달라서 로직을 분리하는 것이라 억지로 분리하고 있다고 느껴지지는 않습니다.
이렇게 분리해두면 향후 새로운 이동 방식을 추가할 때도 유연한 확장이 가능하고요.

junhaa replied:
로직을 분리하는 것이 익숙하지 않았던 것 같아요. 실제 이동 규칙으로 분리하고 보니, 오히려 OCP를 만족하면서 새로운 로직을 쉽게 추가할 수 있는 것 같아요!

pkeugine commented on src/main/java/domain/Movements.java:
```diff
@@ -0,0 +1,54 @@
+                .findIntermediatePositions(startPosition, endPosition);
+    }
+
+    public boolean canMovePieceToPosition(Piece target, Position expectedPosition, List<Piece> alivePieces) {
+        if (canNotMoveTargetTo(target, expectedPosition)) {
+            return false;
+        }
+        if (hasPieceAtIntermediatePositions(target, expectedPosition, alivePieces)) {
+            return false;
+        }
+        return true;
+    }
```
아래와 같이 변경하면 더 간결하고 `canNotMoveTargetTo` 메소드도 필요 없어질 것 같습니다. 

```suggestion
 public boolean canMovePieceToPosition(Piece target, Position expectedPosition, List<Piece> alivePieces) {
    return this.canMoveFromTo(target.getPosition(), expectedPosition) &&
            !hasPieceAtIntermediatePositions(target, expectedPosition, alivePieces);
}
```

junhaa replied:
부정문을 제거하기 위해 `canNotMoveTargetTo` 메소드를 따로 생성했는데, `canMoveFromTo`를 이용하여 작성한 코드가 더 읽기 편한 것 같네요 🥲 `if`분기를 제거하니, 더욱 간결해진 것 같아요. 

위 상황은 아니지만, 조건 분기가 더욱 많아진다면(3~4개의 조건 분기) `&&`를 이용하여 코드를 보는 입장에서 조건을 이해하는 게 어려울 것 같은데, 그런 경우는 따로 if문을 통해 작성하는 게 좋을까요? 물론 지금의 경우에는 피케이가 작성해주신 방법이 더욱 가독성이 좋은 것 같아요!

pkeugine replied:
조건이 길어져서 가독성이 떨어질 경우 해당 조건을 변수로 추출하는 것이 좋습니다.
변수 이름만 보고도 그 조건의 의미를 파악할 수 있게 지정하면 코드를 보는 입장에서도 조건을 더 잘 이해할 수 있으니까요.

junhaa replied:
여러 조건들을 변수로 두어 **이름을 통해 어떤 의도를 가지고 있는 지** 나타내면 이후 내부 조건문을 읽을 때 쉽게 이해할 수 있을 것 같네요. 🤔 답변 감사해요 피케이 😃

pkeugine commented on src/main/java/domain/Movements.java:
```diff
@@ -0,0 +1,54 @@
+        this.movements = movements;
+    }
+
+    public boolean canMoveFromTo(Position startPosition, Position endPosition) {
```
클래스 내부에서만 쓰이는 메소드는 `private`로 구현해주세요.
`canMoveFromTo`, `findIntermediatePositions`

테스트를 위해 `public`으로 열어두신 것 같지만
더 넓은 범위의 `public` 메소드를 테스트하는 것으로 대체할 수 있습니다.

junhaa replied:
TDD로 작성하다가 `public`으로 선언한 후 클래스 내부에서만 쓰이지만 변경을 하지 못했네요 😭 메소드마다 사용되는 위치 확인한 후 클래스 내부에서만 사용되는 메소드는 `private`로 변경하도록 하겠습니다!

pkeugine commented on src/main/java/domain/Movements.java:
```diff
@@ -0,0 +1,54 @@
+
+    public List<Position> findIntermediatePositions(Position startPosition, Position endPosition) {
+        return movements.stream()
+                .filter(movement -> movement.isValidMove(startPosition, endPosition))
```
`canMoveFromTo`, `findIntermediatePositions` 모두 `canMovePieceToPosition` 메소드에서 사용되는 메소드인데,
`movement.isValidMove()` 를 중복해서 호출하고 있습니다.

junhaa replied:
둘 다 같은 메서드를 호출하고 있는데 확인하지 못했네요 😭 `canMoveFromTo`를 검증하는 중복되기 때문에 부분은 제거하도록 하겠습니다!

pkeugine commented on README.md:
```diff
@@ -1,3 +1,86 @@
+루키가 옮길 위치의 좌표를 입력해주세요. 예시 -> a5
+a5
+
+abcdefghi
```
행과 열을 알파벳과 숫자로 헷갈리지 않게 구분해준 것 좋네요 👍

pkeugine reviewed (changes requested):
안녕하세요 루키, 수고하셨습니다.
개선하면 좋을 부분들 코멘트 남겼습니다.
확인해보시고 궁금한 점이 있다면 댓글이나 디엠 남겨주세요 :)
레벨 1 마지막 미션 화이팅입니다!

junhaa committed:
refactor: Movements 내부 로직 추상 클래스로 이동하여 코드 중복 제거

junhaa committed:
refactor: 기물의 이동 경로를 반환하는 기능 분리

junhaa committed:
refactor: 기물의 이동 경로로 이동이 가능한지 검증하는 로직 분리

junhaa committed:
refactor: 기물 별 이동 로직 분리에 따른 테스트 코드 변경

junhaa committed:
refactor: 위치 좌표를 Board에서 관리하도록 변경

junhaa committed:
refactor: Row Column의 공통 로직 추상 클래스로 분리

junhaa committed:
style: 코드 컨벤션 적용

junhaa committed:
refactor: 게임 진행 여부 조건을 JanggiGame에서 호출하여 불필요한 depth 제거

junhaa committed:
refactor: 불변성 보장으로 인한 객체 생성 기능 제거

- 기존 Piece에서는 Position을 변경하기 때문에 불변성 보장 X -> Position을 Board로 이동하여 불변성 보장
- 깊은 복사를 위한 getInstance()의 메서드를 제거

junhaa committed:
feat: 사용자가 잘못된 입력을 한 경우 재입력 받는 기능 구현

junhaa commented:
안녕하세요 피케이 😄  
주말 잘 보내고 계신가요?

피케이께서 남겨주신 피드백을 바탕으로 코드를 수정해봤어요.  

특히 기물의 **이동 로직이 지나치게 복잡**하게 작성되어 있다는 점에 공감했고, 구조를 단순화하는 방향으로 리팩토링을 진행했어요. 🤔

### 주요 개선 방향

피드백을 반영하면서 다음과 같은 기준을 가지고 구조를 다시 고민해봤어요.

1. `Movements` 내부 로직이 너무 복잡하므로, 이를 단순화할 수 있는 구조로 변경할 것
    
2. 기물들이 공통적으로 가지는 로직은 `Piece` 클래스에서 처리해 **중복을 제거**할 것
    
3. 각 기물이 자신만의 `Movements` 기능을 갖도록 하여, **책임을 명확히 분리**할 것
    

### 이동 방식 분류 기준

이동 로직을 구현하는 방식에 대해 다음 두 가지를 고민했는데요, 

1. **고정 거리 이동** vs **가변 거리 이동**
    
    - 예: 졸, 상, 마는 고정 거리 이동
        
    - 차, 포는 가변 거리(막힐 때까지) 이동

2. 직선 이동 vs 직선 + 대각선 이동

    - 예: 상, 마는 직선 + 대각선 이동

    - 나머지 기물들은 모두 직선으로만 이동

두 방식 중에서는 **고정 거리 이동 vs 가변 거리 이동**으로 구분하는 것이 추후 이동 방식 변경 시 수정 범위가 적어 유지보수에 유리할 것 같아 선택했어요.

### 이동 가능 여부 검증 방식

기물은 이동 가능 여부를 검증하는 책임을 두고 있어요. 제가 생각한 체스판에서의 **이동 가능 여부**를 판단하는 기능은 크게 두 가지로 나눌 수 있다고 생각했어요.

1. **해당 기물이 해당 좌표로 이동할 수 있는지** (이동 규칙 검증)
    
2. **이동 경로 상에 문제가 없는지** (경로 상의 장애물, 아군 존재 여부 등)
    

그래서 각 기물이 `MoveRule`과 `PathValidator`를 가지도록 구조화하여,  
**기물 자체가 자신의 이동 가능 여부를 판단**할 수 있도록 설계했어요.


이 외에도 다음과 같은 코드 수정들이 있었어요.

 - 위치 좌표(`Position`)을 `Board`에서 관리하도록 변경
 - `Row`, `Column` 클래스를 공통 추상 클래스(`Axis`)로 분리
 - 재입력 기능 구현

---

이번 미션을 진행하면서, 궁금했던 점들이 있었어요. 🤔

### 1. 읽기 좋은 코드의 메서드 위치

현재 코드에서 **메서드를 어떤 기준으로 두어야 읽기 좋은지**에 대해 고민이 들었어요. 제가 생각하는 방법은 다음과 같아요.
1. 접근 지정자 (`Public` -> `Protected` -> `private`) 순서
2. 메서드가 호출되는 순서, 관련된 코드는 가깝게

현재 두 가지를 적절하게 사용하고 있는데, 피케이는 메서드를 둘 때 어떤 기준이 있는 지 궁금해요!

### 2. 불변성을 보장하지 못하는 객체 처리

이전 코드에서는 `Piece` 자체가 `Position`을 가지고, 이동할 때 마다 `Position`을 변경해 주는 형식으로 활용했어요. 따라서 `Position`이 변경되기 때문에 `Piece`가 **불변성을 보장하지 못했어요.**

출력을 위해서 현재 `Piece`들을 가져오는 메서드가 있었는데요. 객체의 안정성을 보장하기 위해 다음과 같은 2가지 방법이 있었어요.

1. `Piece`의 `Position`을 fianl로 두고, `Piece`가 이동할 때 마다 새로 만들어서 반환한다.
    - 불변 객체가 되기 때문에 사이드 이펙트가 줄어든다.
    - 코드를 사용하는 입장에서는 반환되는 객체가 어떤 것인지 알기 힘들다.
    - 빈번한 이동이 있는 경우 새 객체를 생성하기 때문에 많은 자원이 필요하다.

2. `Piece`들을 반환할 때, 깊은 복사를 통해 반환한다.
    - 기존 객체를 계속 사용할 수 있다.
    - 상태 공유가 있는 경우 사이드 이펙트가 발생할 수 있다.

위 두 가지 상황에서 1번 방법의 단점이 크다고 생각해서, 결국 2번 방법을 선택했는데, 피케이는 어떻게 생각하는지 궁금해요! 

---

아직 코드에서 수정해야 할 부분이 많은 것 같아요.

수정해야 할 부분에 피드백 남겨주시면 최대한 빨리 고쳐볼게요! 😄

junhaa committed:
fix: 게임이 종료되어도 계속 진행되는 오류 해결

pkeugine commented on src/main/java/domain/Board.java:
```diff
@@ -0,0 +1,84 @@
+                .filter(piece -> piece.isSameType(PieceType.KING))
+                .count();
+
+        return kingCount == GAME_RUNNING_KING_COUNT;
```
커밋 중에 커밋 메세지에 맞지 않은 내용이 포함되어 있는게 좀 있습니다.
커밋 단위로 코드를 보며 리뷰 할 때도 많고,
더 나아가 애플리케이션 버전 릴리즈나 롤백 할 때도 꽤 중요해질 때가 있으니 앞으로 잘 관리해보면 좋겠네요. :)

junhaa replied:
이 부분에 대해서는 저도 많이 부족하다고 생각합니다 😭 아무래도 커밋 자체를 미루다가 모두 완성하고 한꺼번에 커밋 하려다 보니 내용에 맞지 않는 부분이 포함되기도 하는 것 같아요. 

> 커밋 단위로 코드를 보며 리뷰 할 때도 많고,
더 나아가 애플리케이션 버전 릴리즈나 롤백 할 때도 꽤 중요해질 때가 있으니 앞으로 잘 관리해보면 좋겠네요. :)

피케이가 말씀해주신 부분을 보니 얼른 고쳐야겠다는 생각이 드네요 🥲
최대한 기능 단위를 나누어서 커밋하는 연습을 해보겠습니다!

pkeugine commented on src/main/java/domain/position/Direction.java:
```diff
@@ -0,0 +1,39 @@
+    public boolean canMoveFrom(Position from){
+        return from.canMovePosition(this.deltaRow, this.deltaColumn);
+    }
+
```
불필요한 줄바꿈이 있습니다.

pkeugine commented on src/main/java/domain/position/Direction.java:
```diff
@@ -0,0 +1,39 @@
+        return deltaColumn;
+    }
+
+    public Position moveFrom(Position from){
```
위치를 이동시키는 메소드가 방향 클래스에 있다는게 좀 어색해 보입니다.
클래스의 역할에 대해 고민해보면 좋을 것 같습니다.

junhaa replied:
방향에게 '위치를 이동해라'라는 메소드를 두는 것 보다, 위치에게 '방향으로 이동해라'라는 것이 조금 더 자연스러워 보이네요. 수정하도록 하겠습니다!

pkeugine commented on src/main/java/domain/Players.java:
`Player`, `Players` 관련 테스트도 추가해주세요.
만약 필요없다고 판단된다면 패스하셔도 좋습니다.

junhaa replied:
아직 테스트가 없는 부분이 많아서, 우선 전반적으로 테스트 코드를 모두 작성한 후, 필요 없다고 생각되는 부분을 제거하도록 하겠습니다!

pkeugine commented on src/main/java/domain/piece/path/PathValidator.java:
```diff
@@ -0,0 +1,17 @@
+        validateDestination(to, board, movePiece);
+    }
+
+    void validateMovePath(List<Position> positions, Board board);
```
default 메서드를 활용하신다면 `DefaultPathValidator` 에 있는 로직 역시 비슷한 방법으로 구현할 수 있을 듯 합니다.
클래스 추출, default 메서드 활용이라는 두 방식으로 구분하게 된 이유가 궁금합니다.

junhaa replied:
사실 이번 PathValidator에서 `default` 메소드를 이용한 이유가 없었던 것 같아요. 🥲

그래서 먼저 `DefaultPathValidator`의 메소드를 변경해 보았어요. fbd1979bf2de1e354b69d9c61e396a6d78f7e7bc

변경 이후, `default` 메서드를 사용하면서 생길 수 있는 장단점에 대해 고민해 보았는데요.
제가 생각한 단점은 다음 2가지가 있었어요.

- 인터페이스의 설계에 대한 책임을 넘어서 인터페이스가 구현까지 하게 됨
	- 고수준의 모듈이 인터페이스에서 작성된 default 메소드의 구현에 의존하게 됨
-  Mock/Test 시, 구현에 대한 은닉이 어려움
	- 테스트 시 테스트 대상이 인터페이스의 default 메소드 내부 구현에 의존하게 됨 
	
현재 상황에서는 위의 단점들과 `PathValidator`의 구현체를 가져야 하기 때문에 `default` 메소드의 장점이 없다고 생각하여, 기존 `default` 메소드를 제거한 후 각각의 구현체로 이동시켰습니다.

pkeugine commented on README.md:
> 1. 읽기 좋은 코드의 메서드 위치

개인적으로 저는 `1. 접근 지정자 (Public -> Protected -> private) 순서` 를 선호합니다.
먼저 확인해야 하는, 다른 곳에서 활용될 수 있는 메소드가 맨 위에 와서 규모가 커질 수록 코드를 이해하기가 쉬워집니다.

물론 추후 협업을 할 때는 프로젝트 내의 규칙에  맞춰서 정하는 것이 가장 좋습니다.
`진리의 팀바팀`이 여기 적용되겠네요. :)

pkeugine replied:
> 2. 불변성을 보장하지 못하는 객체 처리

`Piece` 들을 반환하는 것이 어떤 부분을 말씀 하시는 건지 정확히 이해하지 못했습니다.
해당 부분에 코멘트를 달아주시면 다시 답변 드리겠습니다.

junhaa replied:
> 1. 읽기 좋은 코드의 메서드 위치

제가 **규모가 커지는 상황**을 고려하지 못했던 것 같아요. 만약 메서드가 호출되는 순서에 맞춰 관련된 코드를 가깝게 두는 경우 `public`에 대한 메서드를 찾기 힘들 것 같아요.

> 2. 불변성을 보장하지 못하는 객체 처리

이 부분은 제가 봐도 설명이 부족했던 것 같네요 😭

우선 이전의 코드에서는 다음과 같이 `Piece`를 작성했어요.

```java
public class Piece {
	private Position position; 

	public void movePiece(Position movePosition){
		this.position = movePosition; // 불변성을 보장하지 못함 (외부에서 Piece.movePiece()를 호출하면 Position 값이 변경)
	}
	
	...
}
```

위의 코드는 불변성을 보장하지 못하기 때문에, `Piece` 객체를 외부로 전달하는 경우 다음과 같은 2가지 방식을 고려했어요.

### 1. `position`을 final로 불변성을 강화한 후, `movePiece`를 호출 할 때, 새로운 `Piece`를 생성하여 반환

```java
public class Piece {
	private final Position position;

	public Piece movePiece(Position movePosition){
		return new Piece(movePosition); // position이 변경 될 때 새로운 Piece를 생성하여 반환
	}
}
```


### 2. `Piece`를 외부로 반환하는 경우, 깊은 복사를 통해 반환

```java
public class Piece {
	private Position position;

	public void movePiece(Position movePosition){
		this.position = position;
	}

	public Piece newInstance(){
		return new Piece(this.position); // 외부로 전달할 때 객체를 깊은 복사하여 반환
	}
}
```

1번의 경우에는 다음과 같은 장단점이 있었고,
 - 객체 자체가 불변성을 가질 수 있다.
 - 사용자의 입장에서는 반환되는 Piece가 어떤 객체인지 알 수 없다.
 - move가 이루어 질 때 마다 새로운 객체를 생성해야 한다.

2번의 경우에도 다음과 같은 장단점들이 있었어요.
 - 객체가 불변성을 가지지 못한다.

위 두 가지 상황 중, `사용자의 입장에서는 반환되는 Piece가 어떤 객체인지 알 수 없다.` 이 단점이 크다고 생각하여 2번의 방법을 선택했는데, 결국 현재 상황에서는 **`view`에 출력하기 위해 `Piece`에 대해 깊은 복사하여 반환하는 메서드**가 생긴다고 생각했어요. 피케이는 위의 두 상황에 대해 어떻게 생각하시는 지 궁금해요!

pkeugine replied:
현재 `Piece` 코드에는 이 부분이 없기 때문에 그 쓰임에 대해서 정확히 파악하기가 어려워요.
제가 잘못 이해하고 답변했다면 말해주세요.

---

각 방법에 장단점이 있습니다.
우선 1번 방법은 `movePiece` 메소드 실행 시 현재의 객체를 변경하지 않고 신규 객체를 리턴합니다. 
이 부분을 혼동하지 않도록 잘 구현해야 할 것 같습니다. 
기존 객체가 변경 됐다고 착각하고 사용할 수 있으니까요.

2번 방법은 실제 객체를 변경하여 그런 혼동은 없을 것 같은데, `newInstance` 라는 이름이 어색합니다.
보통 `newInstance` 라고 하면 초기화된 객체가 리턴될 거라고 기대하거든요.
깊은 복사임을 나타내는 메소드 이름으로 변경하면 더 명확해질 것 같네요. :)

pkeugine reviewed (changes requested):
안녕하세요 루키,
추가 코멘트 및 질문에 대한 답변 남겼습니다.
한 번 확인해주시고, 변경할 내용이 있다면 수정해주세요.

junhaa committed:
style: 불필요한 줄바꿈 제거

junhaa committed:
refactor: 방향에 대한 위치 이동 기능 Position에서 관리하도록 변경

junhaa committed:
test: Path 이동 관련 테스트 코드 추가

junhaa committed:
test: 본인 말 이동 검증 테스트 추가

junhaa committed:
test: 방향 이동 로직 테스트 코드 작성

junhaa committed:
test: 상차림 옵션 생성 기능 테스트 코드 구현

junhaa committed:
test: Usernames 이름 확인 테스트 종류 추가

junhaa committed:
test: Player 사용자 테스트 코드 작성

junhaa committed:
test: Players 팀별 사용자에 대한 테스트 코드 작성

junhaa committed:
refactor: default 메소드를 이용하여 DefaultPathValidator 내부 검증 로직 인터페이스로 이동

junhaa committed:
refactor: PathValidator의 default 메소드 제거 후 DefaultPathValidator 검증 로직 이동

- default 메소드의 단점 (테스트, DIP)를 고려하여 로직 처리 위치 변경

junhaa commented:
안녕하세요 피케이 😃

피드백 주신 내용을 읽고, 코드를 수정해봤어요.

이번 수정에서 커밋을 신경쓰면서 해봤어요. 🥲

혹시라도 커밋 방법에서 잘못된 부분이 있다면 피드백 부탁드려요 😀

pkeugine commented on src/main/java/domain/piece/path/CannonPathValidator.java:
```diff
@@ -0,0 +1,65 @@
+
+    }
+
+    public void validateDestination(Position to, Board board, Piece movePiece) {
```
테스트 만을 위해 `public` 으로 열려있는 메소드 같네요.
`public` 메소드인 `validatePath` 테스트를 통해서 `validateDestination`, `validateMovePath` 테스트 모두 가능할 것 같습니다.
테스트만을 위해 외부 사용 가능성을 열어두는 것은 지양하는 것이 좋습니다.

junhaa replied:
기존의 `default` 메서드를 이용할 때 사용한 인터페이스 메서드를 제거하면서 접근지정자를 변경하지 못했던 것 같아요. `private`로 변경 후 테스트 코드도 변경하도록 하겠습니다!

pkeugine commented on src/test/java/domain/piece/path/CannonPathValidatorTest.java:
```diff
@@ -0,0 +1,104 @@
+        // given
+        CannonPathValidator cannonPathValidator = new CannonPathValidator();
+        List<Position> positions = List.of(Position.of(3, 3), Position.of(3, 4), Position.of(3, 5));
+        Cannon cannon = new Cannon(TeamType.HAN);
```
사용하지 않는 변수가 있습니다.

pkeugine commented on src/main/java/.gitkeep:
`.gitkeep` 파일은 더 이상 사용하지 않으니 제거해주세요.

pkeugine commented on src/test/java/domain/piece/strategy/InnerElephantStrategyTest.java:
```diff
@@ -0,0 +1,29 @@
+
+        // then
+        Position elephantPosition = Position.of(0, 2);
+        assertThat(choElephant.get(elephantPosition).isSameType(PieceType.ELEPHANT));
```
`assertThat` 이후 아무 검사를 하고 있지 않습니다. 
아래처럼 `isTrue()`까지 호출해야 의도된 검사가 됩니다.
다른 `StrategyTest` 들도 모두 점검해주세요 :)

```suggestion
        assertThat(choElephant.get(elephantPosition).isSameType(PieceType.ELEPHANT)).isTrue();
```

pkeugine commented on src/test/java/domain/piece/strategy/InnerElephantStrategyTest.java:
```diff
@@ -0,0 +1,29 @@
+        Position elephantPosition = Position.of(0, 2);
+        assertThat(choElephant.get(elephantPosition).isSameType(PieceType.ELEPHANT));
+    }
+
```
이 부분 공백 유무가 `StrategyTest` 별로 다르네요.
한 번 통일시켜볼까요?

pkeugine commented on src/test/java/domain/BoardTest.java:
```diff
@@ -0,0 +1,157 @@
+
+        // then
+        Map<Position, Piece> alivePieces = board.getAlivePieces();
+        assertThat(alivePieces.containsKey(endPosition)).isTrue();
```
```suggestion
        assertThat(alivePieces).containsKey(endPosition);
```
위와 같이 변경해도 동일합니다.

pkeugine reviewed (approved):
안녕하세요 루키,
빠르게 발전하는 모습이 보이네요 👍 
추가 코멘트 남겼으니 한 번 확인 부탁드립니다.
우선 이번 단계는 여기서 머지하고, 다음 단계에서 계속 진행해보죠.
마지막까지 파이팅입니다 💪 💪 💪
